File: patterns/concurrency/active-object.py
"""
Active Object Pattern - Python

Encapsulates method execution in its own thread of control,
decoupling method invocation from method execution.
"""

import threading
import queue
import time


class ActiveObject:
    def __init__(self):
        self.queue = queue.Queue()
        self.running = True
        self.thread = threading.Thread(target=self._run)
        self.thread.start()
    
    def _run(self):
        while self.running:
            try:
                method, args, kwargs, future = self.queue.get(timeout=0.1)
                try:
                    result = method(*args, **kwargs)
                    future.set_result(result)
                except Exception as e:
                    future.set_exception(e)
            except queue.Empty:
                continue
    
    def execute(self, method, *args, **kwargs):
        future = Future()
        self.queue.put((method, args, kwargs, future))
        return future
    
    def add(self, a, b):
        time.sleep(0.1)
        return a + b
    
    def multiply(self, a, b):
        time.sleep(0.1)
        return a * b
    
    def stop(self):
        self.running = False
        self.thread.join()


class Future:
    def __init__(self):
        self._result = None
        self._exception = None
        self._condition = threading.Condition()
        self._done = False
    
    def set_result(self, result):
        with self._condition:
            self._result = result
            self._done = True
            self._condition.notify_all()
    
    def set_exception(self, exception):
        with self._condition:
            self._exception = exception
            self._done = True
            self._condition.notify_all()
    
    def result(self):
        with self._condition:
            while not self._done:
                self._condition.wait()
            if self._exception:
                raise self._exception
            return self._result


# Example usage
if __name__ == "__main__":
    print("=== Active Object Pattern ===\n")
    
    active_object = ActiveObject()
    
    future1 = active_object.execute(active_object.add, 5, 3)
    future2 = active_object.execute(active_object.multiply, 4, 7)
    
    print("Add result:", future1.result())
    print("Multiply result:", future2.result())
    
    active_object.stop()

File: patterns/concurrency/active-object.js
/**
 * Active Object Pattern - JavaScript
 * 
 * Encapsulates method execution in its own thread of control,
 * decoupling method invocation from method execution.
 */

class ActiveObject {
    constructor() {
        this.queue = [];
        this.running = false;
        this.start();
    }
    
    start() {
        this.running = true;
        this.processQueue();
    }
    
    async processQueue() {
        while (this.running) {
            if (this.queue.length > 0) {
                const { method, args, resolve, reject } = this.queue.shift();
                try {
                    const result = await method.apply(this, args);
                    resolve(result);
                } catch (error) {
                    reject(error);
                }
            } else {
                await new Promise(resolve => setTimeout(resolve, 10));
            }
        }
    }
    
    async execute(method, ...args) {
        return new Promise((resolve, reject) => {
            this.queue.push({ method, args, resolve, reject });
        });
    }
    
    async add(a, b) {
        await new Promise(resolve => setTimeout(resolve, 100));
        return a + b;
    }
    
    async multiply(a, b) {
        await new Promise(resolve => setTimeout(resolve, 100));
        return a * b;
    }
    
    stop() {
        this.running = false;
    }
}

// Example usage
console.log('=== Active Object Pattern ===\n');

const activeObject = new ActiveObject();

(async () => {
    const result1 = await activeObject.execute(activeObject.add, 5, 3);
    console.log('Add result:', result1);
    
    const result2 = await activeObject.execute(activeObject.multiply, 4, 7);
    console.log('Multiply result:', result2);
    
    activeObject.stop();
})();

module.exports = { ActiveObject };

